### PR TITLE
Upgrade `pylint` and `mypy` in project configuration

### PR DESCRIPTION
**Pull request by vijaypavann-db**
* Plugins
   * Added 
         *  `databricks-labs-pylint`
         *  `mypy`
    * Removed
         *  `isort`

### CODE REVIEW

File: .github/workflows/push.yml [hunk 1]
```diff
@@ -61,8 +61,13 @@ jobs:
       - name: Install hatch
         run: pip install hatch==$HATCH_VERSION
 
-      - name: Verify linting
-        run: hatch run verify
+      - name: Reformat code
```
Comment (sundarshankar89) [inline .github/workflows/push.yml:64]: Why do we need this here?

**Review by sundarshankar89 — CHANGES_REQUESTED** [review]

File: src/databricks/labs/remorph/coverage/remorph_snow_transpilation_coverage.py [hunk 2]
```diff
@@ -6,8 +6,8 @@
 from databricks.labs.remorph.snow.snowflake import Snow
 
 if __name__ == "__main__":
-    input_dir = commons.get_env_var("INPUT_DIR", required=True)
-    output_dir = commons.get_env_var("OUTPUT_DIR", required=True)
+    input_dir = commons.get_env_var("INPUT_DIR", required=True) or ""
```
Comment (sundarshankar89) [inline src/databricks/labs/remorph/coverage/remorph_snow_transpilation_coverage.py:9]: This is wrong, we need to update the rule, to ignore this.

**Review by sundarshankar89 — CHANGES_REQUESTED** [review]

File: pyproject.toml [hunk 3]
```diff
@@ -49,21 +49,24 @@ dependencies = [
   "pytest-cov>=4.0.0,<5.0.0",
   "black>=23.1.0",
   "ruff>=0.0.243",
-  "isort>=2.5.0",
-  "databricks-connect"
+  "databricks-connect",
+  "types-pyYAML",
+  "types-pytz",
+  "databricks-labs-pylint~=0.1.0",
```
Comment (nfx) [inline pyproject.toml:55]:
```suggestion
  "databricks-labs-pylint~=0.4.0",
```
we have newer versions :)

**Review by nfx — CHANGES_REQUESTED** [review]

File: .github/workflows/push.yml (hunk 1 above)
Comment (vijaypavann-db) [reply .github/workflows/push.yml:66]: Following the same flow as [ucx](https://github.com/databrickslabs/ucx/blob/main/.github/workflows/push.yml)

File: src/databricks/labs/remorph/intermediate/dag.py [hunk 4]
```diff
@@ -6,8 +6,8 @@
 class Node:
     def __init__(self, name: str):
         self.name = name.lower()
-        self.children = []
-        self.parents = []
+        self.children: list[str] = []
+        self.parents: list[str] = []
 
     def add_parent(self, node: str):
```
Comment (sundarshankar89) [inline src/databricks/labs/remorph/intermediate/dag.py:12]: Missing return types for functions in entire file.

File: src/databricks/labs/remorph/reconcile/compare.py [hunk 5]
```diff
@@ -17,9 +17,9 @@ def raise_column_mismatch_exception(msg: str, source_missing: list[str], target_
 
 def reconcile_data(source: DataFrame, target: DataFrame, key_columns: list[str], report_type: str) -> ReconcileOutput:
     source_alias = "src"
-    target_alias = "tgt"
+    target_alias: str = "tgt"
```
Comment (sundarshankar89) [inline src/databricks/labs/remorph/reconcile/compare.py:20]: why is this needed here?

File: src/databricks/labs/remorph/reconcile/query_builder/expression_generator.py [hunk 6]
```diff
@@ -53,7 +53,7 @@ def sort_array(expr: exp.Expression, asc=True):
     return _apply_func_expr(expr, exp.SortArray, asc=exp.Boolean(this=asc))
 
 
-def to_char(expr: exp.Expression, to_format=None, nls_param=None):
+def to_char(expr: exp.Expression, to_format=None, nls_param=None) -> exp.ToChar | exp.Expression:
```
Comment (sundarshankar89) [inline src/databricks/labs/remorph/reconcile/query_builder/expression_generator.py:56]:
```suggestion
def to_char(expr: exp.Expression, to_format=None, nls_param=None) -> exp.Expression:
```

Comment (sundarshankar89) [reply src/databricks/labs/remorph/reconcile/query_builder/expression_generator.py:56]: applies for entire file

File: src/databricks/labs/remorph/snow/experimental.py [hunk 7]
```diff
@@ -1,24 +1,22 @@
 import logging
-from typing import ClassVar
 
 from sqlglot import expressions as exp
 
-from databricks.labs.remorph.snow import databricks
+from databricks.labs.remorph.snow import databricks as custom_databricks
```
Comment (sundarshankar89) [inline src/databricks/labs/remorph/snow/experimental.py:5]:
```suggestion
from databricks.labs.remorph.snow import databricks as extended_databricks
```

File: src/databricks/labs/remorph/snow/lca_utils.py [hunk 8]
```diff
@@ -13,15 +14,15 @@
...
+    dialect: DialectType,
     sql: str,
     filename: str,
 ) -> ValidationError | None:
     """
     Check for presence of unsupported lateral column aliases in window expressions and where clauses
     :return: An error if found
     """
-    if dialect.upper() == "SNOWFLAKE":
+    if str(dialect).upper() == "SNOWFLAKE":
```
Comment (sundarshankar89) [inline src/databricks/labs/remorph/snow/lca_utils.py:25]: Refactor this to use `SQLGLOT_DIALECTS` from `config.py`

File: src/databricks/labs/remorph/snow/lca_utils.py [hunk 9]
```diff
@@ -56,12 +57,17 @@ def check_for_unsupported_lca(
...
+    if root_select is None:
+        raise ValueError("Failed to build scope for the select expression")
+
     # We won't search inside nested selects, they will be visited separately
     nested_selects = {*root_select.derived_tables, *root_select.subqueries}
     alias_info = _find_aliases_in_select(expr)
-    where_ast: Expression = expr.args.get("where")
-    if where_ast:
+    where_ast = expr.args.get("where")
+    if where_ast and isinstance(where_ast, Expression):
```
Comment (sundarshankar89) [inline src/databricks/labs/remorph/snow/lca_utils.py:70]: Do we really need this checks?

**Review by sundarshankar89 — CHANGES_REQUESTED** [review]

File: src/databricks/labs/remorph/helpers/file_utils.py [hunk 10]
```diff
@@ -67,16 +67,18 @@ def dir_walk(root: Path):
...
     :param input_path: Path
     :return: List of SQL files
     """
+    filenames = []
     for _, _, files in dir_walk(Path(input_path)):
         for filename in files:
             if is_sql_file(filename):
-                yield filename
+                filenames.append(filename)
+    return filenames
```
Comment (bishwajit-db) [inline src/databricks/labs/remorph/helpers/file_utils.py:81]: Why we are not using yield? We could have used type Generator[Path, None, None]. If it is required to remove yield, we should update the comments also.

**Review by bishwajit-db — CHANGES_REQUESTED** [review]

File: src/databricks/labs/remorph/coverage/remorph_snow_transpilation_coverage.py (hunk 2 above)
Comment (vijaypavann-db) [reply src/databricks/labs/remorph/coverage/remorph_snow_transpilation_coverage.py:9]: Done

File: src/databricks/labs/remorph/coverage/sqlglot_snow_transpilation_coverage.py [hunk 11]
```diff
@@ -7,8 +7,8 @@
 from databricks.labs.remorph.coverage import commons
 
 if __name__ == "__main__":
-    input_dir = commons.get_env_var("INPUT_DIR", required=True)
-    output_dir = commons.get_env_var("OUTPUT_DIR", required=True)
+    input_dir: str = commons.get_env_var("INPUT_DIR", required=True) or ""
```
[the remark this answers is not in the document]
Comment (vijaypavann-db) [reply src/databricks/labs/remorph/coverage/sqlglot_snow_transpilation_coverage.py:10]: Done

File: src/databricks/labs/remorph/helpers/file_utils.py (hunk 10 above)
Comment (vijaypavann-db) [reply src/databricks/labs/remorph/helpers/file_utils.py:81]: yield is returning the `Generator` which is not the expected return type. 

`error:The return type of a generator function should be "Generator"or one of its supertypes `
`for filename in get_sql_file(self.input_path):` 

the above line expects only `Path`, not `Generator Type` so changed it to `list[Path]`. 

Will update the Doc string.

File: src/databricks/labs/remorph/intermediate/dag.py (hunk 4 above)
Comment (vijaypavann-db) [reply src/databricks/labs/remorph/intermediate/dag.py:12]: Done

File: src/databricks/labs/remorph/reconcile/query_builder/expression_generator.py (hunk 6 above)
Comment (vijaypavann-db) [reply src/databricks/labs/remorph/reconcile/query_builder/expression_generator.py:56]: Done

File: src/databricks/labs/remorph/snow/lca_utils.py (hunk 8 above)
Comment (vijaypavann-db) [reply src/databricks/labs/remorph/snow/lca_utils.py:25]: Done

File: src/databricks/labs/remorph/snow/lca_utils.py (hunk 9 above)
Comment (vijaypavann-db) [reply src/databricks/labs/remorph/snow/lca_utils.py:67]: Since it could be Expression | None, we are checking for this.

File: pyproject.toml (hunk 3 above)
Comment (vijaypavann-db) [reply pyproject.toml]: Updated the version

**Review by sundarshankar89 — CHANGES_REQUESTED** [review]

File: src/databricks/labs/remorph/reconcile/compare.py (hunk 5 above)
Comment (vijaypavann-db) [reply src/databricks/labs/remorph/reconcile/compare.py]: Removed

File: src/databricks/labs/remorph/coverage/remorph_snow_transpilation_coverage.py [hunk 12]
```diff
@@ -19,6 +19,6 @@
         remorph_version,
         Snow,
         Databricks,
-        Path(input_dir),
-        Path(output_dir),
+        Path(str(input_dir)),
```
[the remark this answers is not in the document]
Comment (vijaypavann-db) [reply src/databricks/labs/remorph/coverage/remorph_snow_transpilation_coverage.py:22]:
```
input_dir = input_dir if input_dir is not None else ""
output_dir = output_dir if output_dir is not None else ""
```

Modified a bit but `os.getenv(env_var)` `returns str | None`, so need to handle it before calling the function.

File: src/databricks/labs/remorph/coverage/sqlglot_snow_transpilation_coverage.py [hunk 13]
```diff
@@ -18,6 +18,6 @@
         sqlglot_version,
         Snowflake,
         Databricks,
-        Path(input_dir),
-        Path(output_dir),
+        Path(str(input_dir)),
```
[the remark this answers is not in the document]
Comment (vijaypavann-db) [reply src/databricks/labs/remorph/coverage/sqlglot_snow_transpilation_coverage.py:21]: `os.getenv(env_var)` returns `str | None`, so need to handle it before calling the function.

File: src/databricks/labs/remorph/snow/sql_transpiler.py [hunk 14]
```diff
@@ -48,3 +48,4 @@ def parse_sql_content(self, sql, file_name):
     def _find_root_tables(expression) -> str:
         for table in expression.find_all(exp.Table, bfs=False):
             return table.name
+        return ""
```
[the remark this answers is not in the document]
Comment (vijaypavann-db) [reply src/databricks/labs/remorph/snow/sql_transpiler.py:52]: Done

File: src/databricks/labs/remorph/transpiler/execute.py [hunk 15]
```diff
@@ -120,8 +121,8 @@ def _process_directory(
     return counter, parse_error_list, validate_error_list
 
 
-def _process_recursive_dirs(config: MorphConfig, validator: Validator, transpiler: SqlglotEngine):
-    input_sql = Path(config.input_sql)
+def _process_recursive_dirs(config: MorphConfig, validator: Validator | None, transpiler: SqlglotEngine):
+    input_sql = Path(str(config.input_sql))
```
[the remark this answers is not in the document]
Comment (vijaypavann-db) [reply src/databricks/labs/remorph/transpiler/execute.py]:
```
input_sql_path = config.input_sql if config.input_sql else ""
```
Since it is returning str | None, so handled it using the above code

File: src/databricks/labs/remorph/coverage/remorph_snow_transpilation_coverage.py [hunk 16]
```diff
@@ -9,6 +9,10 @@
     input_dir = commons.get_env_var("INPUT_DIR", required=True)
     output_dir = commons.get_env_var("OUTPUT_DIR", required=True)
 
+    # Ensure input_dir and output_dir are not None
+    input_dir = input_dir if input_dir is not None else ""
```
Comment (sundarshankar89) [inline src/databricks/labs/remorph/coverage/remorph_snow_transpilation_coverage.py:13]: Do error handling here, instead of pushing the input as string.

File: src/databricks/labs/remorph/coverage/sqlglot_snow_transpilation_coverage.py [hunk 17]
```diff
@@ -12,6 +12,10 @@
     sqlglot_version = sqlglot.__version__
     SQLGLOT_COMMIT_HASH = ""  # C0103 pylint
 
+    # Ensure input_dir and output_dir are not None
+    input_dir = input_dir if input_dir is not None else ""
```
Comment (sundarshankar89) [inline src/databricks/labs/remorph/coverage/sqlglot_snow_transpilation_coverage.py:16]: Same as above

File: src/databricks/labs/remorph/snow/sql_transpiler.py [hunk 18]
```diff
@@ -22,13 +23,13 @@ def transpile(
 
         return TranspilationResult(transpiled_sql, error_list)
 
-    def parse(self, sql: str, file_name: str) -> (list[Expression] | None, ParserError | None):
+    def parse(self, sql: str, file_name: str) -> tuple[list[Expression | None] | None, ParserError | None]:
```
Comment (sundarshankar89) [inline src/databricks/labs/remorph/snow/sql_transpiler.py:26]:
```suggestion
    def parse(self, sql: str, file_name: str) -> tuple[list[Expression] | None, ParserError | None]:
```

**Review by sundarshankar89 — CHANGES_REQUESTED** [review]

File: src/databricks/labs/remorph/snow/sql_transpiler.py (hunk 18 above)
Comment (vijaypavann-db) [reply src/databricks/labs/remorph/snow/sql_transpiler.py:26]: `sqlglot` `parse` returns `list[Expression | None]`, so couldn't change this.

```
 returns parse(
    sql: str, read: DialectType = None, dialect: DialectType = None, **opts
) -> t.List[t.Optional[Expression]]
```

File: src/databricks/labs/remorph/coverage/sqlglot_snow_transpilation_coverage.py (hunk 17 above)
Comment (vijaypavann-db) [reply src/databricks/labs/remorph/coverage/sqlglot_snow_transpilation_coverage.py]:
```
if not input_dir:
        raise ValueError("Environment variable `INPUT_DIR` is required")
    if not output_dir:
        raise ValueError("Environment variable `OUTPUT_DIR` is required")
```

Handled in this way

File: src/databricks/labs/remorph/coverage/remorph_snow_transpilation_coverage.py (hunk 16 above)
Comment (vijaypavann-db) [reply src/databricks/labs/remorph/coverage/remorph_snow_transpilation_coverage.py]:
```
if not input_dir:
        raise ValueError("Environment variable `INPUT_DIR` is required")
    if not output_dir:
        raise ValueError("Environment variable `OUTPUT_DIR` is required")
```

Handled in this way

**Review by sundarshankar89 — APPROVED** [review]
LGTM

@bishwajit-db Can you take a pass as welll.

File: src/databricks/labs/remorph/transpiler/execute.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (vijaypavann-db) [reply src/databricks/labs/remorph/transpiler/execute.py]: Handled this way: 

```
if not config.input_sql:
        logger.error("Input SQL path is not provided.")
        raise ValueError("Input SQL path is not provided.")

    input_sql = Path(config.input_sql)
```

File: src/databricks/labs/remorph/reconcile/query_builder/base.py [hunk 19]
```diff
@@ -87,31 +88,37 @@ def _user_transformer(node: exp.Expression, user_transformations: dict[str, str]
...
     def _default_transformer(node: exp.Expression, schema: list[Schema], source: Dialect) -> exp.Expression:
+
         def _get_transform(datatype: str):
-            if DataType_transform_mapping.get(source) is not None:
-                if DataType_transform_mapping.get(source).get(datatype.upper()) is not None:
-                    return DataType_transform_mapping.get(source).get(datatype.upper())
-                if DataType_transform_mapping.get(source).get("default") is not None:
-                    return DataType_transform_mapping.get(source).get("default")
-            return DataType_transform_mapping.get("default")
+            source_dialects = [source_key for source_key, dialect in SQLGLOT_DIALECTS.items() if dialect == source]
```
Comment (sundarshankar89) [inline src/databricks/labs/remorph/reconcile/query_builder/base.py:106]: Love this change

File: src/databricks/labs/remorph/reconcile/query_builder/expression_generator.py [hunk 20]
```diff
@@ -214,10 +214,9 @@ def build_between(this: exp.Expression, low: exp.Expression, high: exp.Expressio
     return exp.Between(this=this, low=low, high=high)
 
 
-# [TODO] NEED TO UPDATE THE KEY TO HANDLE DIALECT TYPE
-DataType_transform_mapping = {
-    "default": [partial(coalesce, default='', is_string=True), trim],
-    "snowflake": {exp.DataType.Type.ARRAY.value: [array_to_string, array_sort]},
+DataType_transform_mapping: dict[str, dict[str, list[partial[exp.Expression]]]] = {
+    "default": {"default": [partial(coalesce, default='', is_string=True), partial(trim)]},
```
Comment (sundarshankar89) [inline src/databricks/labs/remorph/reconcile/query_builder/expression_generator.py:218]: nit:
```suggestion
    "universal": {"default": [partial(coalesce, default='', is_string=True), partial(trim)]},
```

File: src/databricks/labs/remorph/transpiler/execute.py [hunk 21]
```diff
@@ -50,7 +50,7 @@ def _process_file(
     with input_file.open("r") as f:
         sql = remove_bom(f.read())
 
-    lca_error = lca_utils.check_for_unsupported_lca(config.source, sql, str(input_file))
+    lca_error = lca_utils.check_for_unsupported_lca(SQLGLOT_DIALECTS.get(config.source.lower()), sql, str(input_file))
```
Comment (sundarshankar89) [inline src/databricks/labs/remorph/transpiler/execute.py:53]: refactor to `get_dialect` from `config.py` since this is no longer private

File: src/databricks/labs/remorph/cli.py [hunk 22]
```diff
@@ -28,7 +28,7 @@ def transpile(
     w: WorkspaceClient,
     source: str,
     input_sql: str,
-    output_folder: str,
+    output_folder: str | None,
```
Comment (vijaypavann-db) [inline src/databricks/labs/remorph/cli.py:31]: `default_config.output_folder` is of type "str | None": 
Error: Incompatible types in assignment expression has type "str | None", variable has type "str"
Fix: updated the type of `output_folder`

**Review by sundarshankar89 — CHANGES_REQUESTED** [review]
Small nit:

File: src/databricks/labs/remorph/config.py [hunk 23]
```diff
@@ -1,15 +1,15 @@
...
+from sqlglot.dialects.dialect import Dialect, Dialects, DialectType
 
 from databricks.labs.remorph.helpers.morph_status import ParserError
 from databricks.labs.remorph.reconcile.recon_config import Table
 from databricks.labs.remorph.snow import databricks, experimental, oracle, snowflake
 
 logger = logging.getLogger(__name__)
 
-SQLGLOT_DIALECTS = {
+SQLGLOT_DIALECTS: dict[str, DialectType] = {
```
Comment (vijaypavann-db) [inline src/databricks/labs/remorph/config.py:12]: `DialectType = t.Union[str, Dialect, t.Type[Dialect], None]`

File: src/databricks/labs/remorph/coverage/commons.py [hunk 24]
```diff
@@ -44,7 +44,7 @@ def write_json_line(file: TextIO, content: ReportEntry):
     file.write("\n")
 
 
-def get_env_var(env_var: str, *, required: bool = False) -> str:
+def get_env_var(env_var: str, *, required: bool = False) -> str | None:
```
Comment (vijaypavann-db) [inline src/databricks/labs/remorph/coverage/commons.py:47]: `os.getenv` returns `str|None`

File: src/databricks/labs/remorph/reconcile/query_builder/expression_generator.py (hunk 20 above)
Comment (vijaypavann-db) [reply src/databricks/labs/remorph/reconcile/query_builder/expression_generator.py]: Done

**Review by sundarshankar89 — APPROVED** [review]
LGTM, fix code cov in this PR and any tech debt to be addressed in separate PR.

**Review by nfx — APPROVED** [review]
lgtm
